3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -248,97 +339,134 @@ object PiiPseudonymizerEnrichment extends ParseableEnrichment {
+  import PiiConstants.ModifiedFields
   implicit val json4sFormats = DefaultFormats
 
-  override def applyStrategy(fieldValue: String): String =
+  override def applyStrategy(fieldValue: String, strategy: PiiStrategy): (String, ModifiedFields) = {
+    val modifiedFields = MutableList[JsonModifiedField]()
     if (fieldValue != null) {
-      compact(render(parse(fieldValue) match {
-        case JObject(jObject) => {
+      (compact(render(parse(fieldValue) match {
+        case JObject(jObject) =>
           val jObjectMap = jObject.toMap
           val updated = jObjectMap.filterKeys(_ == "data").mapValues {
             case JArray(contexts) =>
               JArray(contexts.map {
-                case JObject(context) => modifyObjectIfSchemaMatches(context)
-                case x                => x
+                case JObject(context) =>
+                  val (values, listOfModifiedValues) = modifyObjectIfSchemaMatches(context, strategy)
+                  modifiedFields ++= listOfModifiedValues
+                  values
+                case x => x
               })
-            case JObject(unstructEvent) => modifyObjectIfSchemaMatches(unstructEvent)
-            case x                      => x
+            case JObject(unstructEvent) =>
+              val (values, listOfModifiedValues) = modifyObjectIfSchemaMatches(unstructEvent, strategy)
+              modifiedFields ++= listOfModifiedValues
+              values
+            case x => x
           }
           JObject((jObjectMap ++ updated).toList)
-        }
         case x => x
-      }))
-    } else null
+      })), modifiedFields.toList)
+    } else (null, modifiedFields.toList)
+  }
 
-  private def modifyObjectIfSchemaMatches(context: List[(String, json4s.JValue)]): JObject = {
+  /**
+   * Tests whether the schema for this event matches the schema criterion and if it does modifies it.
+   */
+  private def modifyObjectIfSchemaMatches(context: List[(String, json4s.JValue)],
+                                          strategy: PiiStrategy): (JObject, List[JsonModifiedField]) = {
     val fieldsObj = context.toMap
     (for {
-      schema              <- fieldsObj.get("schema")
-      parsedSchemaMatches <- SchemaKey.parse(schema.extract[String]).map(schemaCriterion.matches).toOption
+      schema <- fieldsObj.get("schema")
+      schemaStr = schema.extract[String]
+      parsedSchemaMatches <- SchemaKey.parse(schemaStr).map(schemaCriterion.matches).toOption
       data                <- fieldsObj.get("data")
       if parsedSchemaMatches
-    } yield JObject(fieldsObj.updated("schema", schema).updated("data", jsonPathReplace(data)).toList)).getOrElse(JObject(context))
+      updated = jsonPathReplace(data, strategy, schemaStr)
+    } yield
+      (JObject(fieldsObj.updated("schema", schema).updated("data", updated._1).toList), updated._2)).getOrElse((JObject(context), List()))
```

Comment:
can we break this down, e.g.:

```scala
(modifiedJson, modifiedFields) = jsonPathReplace()
// ...
yield {
  val updatedFields = fieldsObj
    .updated("schema", schema)
    .updated("data", modifiedJson)
    .toList
  (JObject(updatedFields), modifiedFields))
    .getOrElse((JObject(context), List.empty[JsonModifiedField]))
}
```

you might find better variable names than above